File: frontend/catb.ts
export function createUser(payload: any, db: any) {
  const existing = db.findUserByEmail(payload?.email);

  if (existing) throw new Error("User already exists");

  if (!payload?.password || payload.password.length < 8) {
    throw new Error("Weak password");
  }

  const user = {
    email: payload.email,
    name: payload.name,
  };

  return db.insertUser(user);
}

export function processPayment(amount: number, gateway: any) {
  if (amount <= 0) throw new Error("Invalid amount");

  const res = gateway.charge(amount);

  if (!res || res.status !== "success") {
    throw new Error("Payment failed");
  }

  return res.transactionId;
}

export async function fetchActiveUsers(client: any, retries = 2) {
  let lastError;

  for (let i = 0; i <= retries; i++) {
    try {
      const res = await client.get("/users");
      const users = res.data || [];

      return users.filter((u: any) => u?.is_active && u?.email);
    } catch (e) {
      lastError = e;
    }
  }

  throw new Error(`Failed after retries: ${lastError}`);
}

export function processOrder(order: any, inventory: any, pricing: any) {
  if (!order?.items || !Array.isArray(order.items)) {
    throw new Error("Invalid order");
  }

  let total = 0;

  for (const item of order.items) {
    const stock = inventory.checkStock(item.product_id);

    if (stock < item.quantity) {
      throw new Error(`Out of stock: ${item.product_id}`);
    }

    const price = pricing.getPrice(item.product_id);
    total += price * item.quantity;
  }

  return {
    orderId: order.id,
    total: Math.round(total * 100) / 100,
  };
}

File: backend/catb.py

# function
from typing import Dict

def create_user(payload: Dict, db):
    existing = db.find_user_by_email(payload["email"])
    if existing:
        raise ValueError("User already exists")

    if len(payload["password"]) < 8:
        raise ValueError("Weak password")

    user = {
        "email": payload["email"],
        "name": payload["name"]
    }

    return db.insert_user(user)


#function

def process_payment(amount: float, gateway):
    if amount <= 0:
        raise ValueError("Invalid amount")

    response = gateway.charge(amount)

    if response["status"] != "success":
        raise RuntimeError("Payment failed")

    return response["transaction_id"]



#function
from typing import Dict, List

def fetch_active_users(client, retries: int = 2) -> List[Dict]:
    last_error = None

    for _ in range(retries + 1):
        try:
            response = client.get("/users")
            users = response.json()

            return [
                u for u in users
                if u.get("is_active") and u.get("email")
            ]

        except Exception as e:
            last_error = e

    raise RuntimeError(f"Failed after retries: {last_error}")



#function
def process_order(order, inventory_service, pricing_service):
    total = 0

    for item in order["items"]:
        stock = inventory_service.check_stock(item["product_id"])

        if stock < item["quantity"]:
            raise ValueError(f"Out of stock: {item['product_id']}")

        price = pricing_service.get_price(item["product_id"])
        total += price * item["quantity"]

    return {
        "order_id": order["id"],
        "total": round(total, 2)
    }
